### PR TITLE
Use CorInfoCallConvExtension throughout runtime interop code

### PR DESCRIPTION
**Pull request by jkoritzinsky**
As preparation for implementing #46775, this PR changes the interop code in CoreCLR to use CorInfoCallConvExtension as the representation for calling conventions internally instead of CorPinvokeMap or CorUnmanagedCallingConvention.

This change enables us to only have to add support for propagating new calling conventions significantly more easily since we won't need to determine how to represent new calling conventions using the old enums.

### CODE REVIEW

File: src/coreclr/vm/dllimportcallback.cpp [hunk 1]
```diff
@@ -607,7 +607,6 @@ bool TryGetCallingConventionFromUnmanagedCallersOnly(MethodDesc* pMD, CorInfoCal
     if (nativeCallableInternalData)
     {
         namedArgs[0].InitI4FieldEnum("CallingConvention", "System.Runtime.InteropServices.CallingConvention", (ULONG)(CorPinvokeMap)0);
-        namedArgs[0].InitI4FieldEnum("CallingConvention", "System.Runtime.InteropServices.CallingConvention", (ULONG)(CorPinvokeMap)0);
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/vm/dllimportcallback.cpp:610]: Boo. What is that all about?

File: src/coreclr/vm/dllimportcallback.cpp [hunk 2]
```diff
@@ -643,7 +642,7 @@ bool TryGetCallingConventionFromUnmanagedCallersOnly(MethodDesc* pMD, CorInfoCal
     else
     {
         // Set WinAPI as the default
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/vm/dllimportcallback.cpp:644]: Update comment or remove.

File: src/coreclr/vm/stubgen.cpp [hunk 3]
```diff
@@ -2763,6 +2763,69 @@ void ILStubLinker::SetStubTargetCallingConv(CorCallingConvention uNativeCallingC
...
+            case CorInfoCallConvExtension::Fastcall:
+                m_nativeFnSigBuilder.SetCallingConv(IMAGE_CEE_CS_CALLCONV_FASTCALL);
+                break;
+            default:
+                m_nativeFnSigBuilder.SetCallingConv(IMAGE_CEE_CS_CALLCONV_UNMANAGED);
+                break;
+        }
+    }
+
+    if (m_nativeFnSigBuilder.GetCallingConv() == IMAGE_CEE_CS_CALLCONV_UNMANAGED)
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/vm/stubgen.cpp:2794]: I am leery of this side-effect setting logic. At a minimum add some comments to draw attention to the setting of his in the previous branch. The fact that line 2769 gets this and this subsequent gets a perhaps different value is less than clear.

File: src/coreclr/vm/stubgen.cpp [hunk 4]
```diff
@@ -2763,6 +2763,69 @@ void ILStubLinker::SetStubTargetCallingConv(CorCallingConvention uNativeCallingC
     }
 }
 
+void ILStubLinker::SetStubTargetCallingConv(CorInfoCallConvExtension callConv)
+{
+    LIMITED_METHOD_CONTRACT;
+    CorCallingConvention originalCallingConvention = m_nativeFnSigBuilder.GetCallingConv();
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/vm/stubgen.cpp:2769]: Mark `const`.

File: src/coreclr/vm/dllimport.cpp [hunk 5]
```diff
@@ -2511,6 +2493,31 @@ class DispatchStubState : public StubState // For CLR-to-COM late-bound/eventing
...
+namespace
+{
+    CorInfoCallConvExtension GetCallConvValueForPInvokeCallConv(CorPinvokeMap callConv)
+    {
+        LIMITED_METHOD_CONTRACT;
+
+        switch (callConv)
+        {
+        case 0:
+        case pmCallConvWinapi:
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/vm/dllimport.cpp:2505]: This seems odd. WinAPI is stdcall on Windows. What is this function for?

File: src/coreclr/vm/dllimport.cpp [hunk 6]
```diff
@@ -2938,96 +2945,40 @@ void PInvokeStaticSigInfo::BestGuessNDirectDefaults(MethodDesc* pMD)
...
-{
-#ifdef TARGET_UNIX
-    return pmCallConvCdecl;
-#else // TARGET_UNIX
-    return bIsVarArg ? pmCallConvCdecl : pmCallConvStdcall;
-#endif // !TARGET_UNIX
-}
-
-namespace
+inline CorInfoCallConvExtension GetDefaultCallConv(BOOL bIsVarArg)
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/vm/dllimport.cpp:2948]: `inline` shouldn't needed here.

File: src/coreclr/vm/dllimport.cpp [hunk 7]
```diff
@@ -6708,21 +6659,22 @@ PCODE GetILStubForCalli(VASigCookie *pVASigCookie, MethodDesc *pMD)
...
 
             if (suppressGCTransition)
             {
                 dwStubFlags |= NDIRECTSTUB_FL_SUPPRESSGCTRANSITION;
             }
         }
-
-        if (!TryConvertCallConvValueToPInvokeCallConv(callConv, &unmgdCallConv))
-            COMPlusThrow(kTypeLoadException, IDS_INVALID_PINVOKE_CALLCONV);
+        else
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/vm/dllimport.cpp:6674]: Let's flip the predicate and put this on top.

File: src/coreclr/vm/dllimportcallback.cpp (hunk 1 above)
Comment (jkoritzinsky) [reply src/coreclr/vm/dllimportcallback.cpp:610]: I accidentally duplicated this line in another PR. Just removing the duplicate.

File: src/coreclr/vm/dllimport.cpp (hunk 5 above)
Comment (jkoritzinsky) [reply src/coreclr/vm/dllimport.cpp:2505]: I'll add some comments here. This function converts a CorPinvokeMap calling convention to a CorInfoCallConvExtension calling convention, but we need to know when the user either didn't provide a calling convention or provided WinApi as compared to when the user explicitly provided the platform default calling convention (such as having `CallingConvention = CallingConvention.Stdcall` and running on Windows).

Comment (AaronRobinsonMSFT) [reply src/coreclr/vm/dllimport.cpp:2510]: Sounds good. Comments are warranted and should be sufficient.

File: src/coreclr/vm/dllimport.cpp [hunk 8]
```diff
@@ -2511,6 +2493,36 @@ class DispatchStubState : public StubState // For CLR-to-COM late-bound/eventing
 
 #endif // FEATURE_COMINTEROP
 
+namespace
+{
+    // Use CorInfoCallConvExtension::Managed as a sentinel represent a user-provided WinApi calling convention.
+    CorInfoCallConvExtension CallConvWinApiSentinel = CorInfoCallConvExtension::Managed;
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/vm/dllimport.cpp:2499]: Mark `const`.

**Review by AaronRobinsonMSFT — APPROVED** [review]

File: src/coreclr/vm/dllimport.cpp (hunk 8 above)
Comment (jkoritzinsky) [reply src/coreclr/vm/dllimport.cpp:2499]: - [ ] 
```suggestion
    constexpr CorInfoCallConvExtension CallConvWinApiSentinel = CorInfoCallConvExtension::Managed;
```

**Pull request merged**